File: src/exploration/world-generator.ts
import { createDefaultTileFactoryConfig, type TileTemplate } from "./tile-factory";
import { getHexDistance, type ExploreState, type ExploreTile, type ZoneType } from "./explore-state";
import type { WorldSpec } from "../planning/world-spec-builder";
import { getNpcById, getTileById, listExplorationFlows, listNpcs } from "../planning/content-registry";
import type { ContentExplorationFlow, ContentNpc } from "../planning/content-types";

const START_TILE_KEY = "0,0";
const NPCS_PER_TOWN = 3;

function hashToUnitInterval(input: string): number {
  let hash = 2166136261;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i);
    hash *= 16777619;
  }

  const normalized = (hash >>> 0) / 4294967296;
  return normalized;
}

function applyTemplateToTile(tile: ExploreTile, template: TileTemplate): void {
  tile.zone = template.zone;
  tile.name = template.name;
  tile.description = template.description;
  tile.color = [template.color[0], template.color[1], template.color[2], template.color[3]];
  tile.explorationFlowId = null;
  tile.flowLevel = 0;
  tile.enemyPool = template.enemyPool.map((entry) => ({
    enemyId: entry.enemyId,
    weight: entry.weight,
  }));
  tile.tags = [...template.tags];
  tile.locations = [];
}

function pickCandidateTiles(state: ExploreState, targetZone: ZoneType): ExploreTile[] {
  const pool = state.tiles.filter((tile) => tile.key !== START_TILE_KEY && tile.zone !== targetZone);

  pool.sort((a, b) => {
    const aScore = hashToUnitInterval(`${a.key}:${targetZone}`);
    const bScore = hashToUnitInterval(`${b.key}:${targetZone}`);
    return aScore - bScore;
  });

  return pool;
}

function addTownLocations(tile: ExploreTile, radius: number): void {
  const defaults = createDefaultTileFactoryConfig();
  const buildLocations = defaults.buildLocations;

  if (buildLocations === undefined) {
    return;
  }

  const distanceFromCenter = getHexDistance({ q: 0, r: 0 }, tile.coord);
  const locations = buildLocations("town", {
    key: tile.key,
    coord: { q: tile.coord.q, r: tile.coord.r },
    radius,
    distanceFromCenter,
    roll: hashToUnitInterval(`roll:${tile.key}:${radius}`),
  });

  tile.locations = locations;
}

function ensureZoneMinimums(state: ExploreState, spec: WorldSpec): void {
  const defaults = createDefaultTileFactoryConfig();

  for (const requirement of spec.requiredZones) {
    const currentCount = state.tiles.filter((tile) => tile.zone === requirement.zone).length;
    if (currentCount >= requirement.minCount) {
      continue;
    }

    const needed = requirement.minCount - currentCount;
    const candidates = pickCandidateTiles(state, requirement.zone);

    for (let i = 0; i < needed && i < candidates.length; i += 1) {
      const tile = candidates[i];
      applyTemplateToTile(tile, defaults.templatesByZone[requirement.zone]);
      if (requirement.zone === "town") {
        addTownLocations(tile, state.radius);
      }
    }
  }
}

function pickReplacementZone(tileKey: string, seed: string): ZoneType {
  const pool: ZoneType[] = ["forest", "farmland", "mountain", "ocean"];
  const index = Math.floor(hashToUnitInterval(`${seed}:replace:${tileKey}`) * pool.length);
  return pool[Math.max(0, Math.min(pool.length - 1, index))];
}

function enforceMaxTownTiles(state: ExploreState, spec: WorldSpec): void {
  const townMinRequirement =
    spec.requiredZones.find((entry) => entry.zone === "town")?.minCount ?? 0;
  const maxTownTiles = Math.max(spec.maxTownTiles, townMinRequirement);
  const townTiles = state.tiles.filter((tile) => tile.zone === "town");

  if (townTiles.length <= maxTownTiles) {
    return;
  }

  const protectedKeys: Record<string, true> = {
    [START_TILE_KEY]: true,
  };

  const candidates = townTiles.filter((tile) => protectedKeys[tile.key] !== true);
  candidates.sort((a, b) => hashToUnitInterval(`${spec.seed}:${a.key}`) - hashToUnitInterval(`${spec.seed}:${b.key}`));

  let toConvert = townTiles.length - maxTownTiles;
  for (const tile of candidates) {
    if (toConvert <= 0) {
      break;
    }

    const replacementZone = pickReplacementZone(tile.key, spec.seed);
    const defaults = createDefaultTileFactoryConfig();
    applyTemplateToTile(tile, defaults.templatesByZone[replacementZone]);
    toConvert -= 1;
  }
}

function ensureSpecialTiles(state: ExploreState, spec: WorldSpec): void {
  const defaults = createDefaultTileFactoryConfig();

  for (const special of spec.requiredSpecialTiles) {
    const existing = state.tiles.filter((tile) => tile.metadata.specialTileId === special.id);
    if (existing.length >= special.minCount) {
      continue;
    }

    const preferredZone = special.preferredZone;
    const candidatePool = state.tiles.filter((tile) => {
      if (tile.key === START_TILE_KEY) {
        return false;
      }

      if (tile.metadata.specialTileId !== undefined) {
        return false;
      }

      if (preferredZone !== undefined) {
        return tile.zone === preferredZone;
      }

      return true;
    });

    if (candidatePool.length === 0) {
      continue;
    }

    candidatePool.sort((a, b) => hashToUnitInterval(a.key) - hashToUnitInterval(b.key));
    const selected = candidatePool[0];

    if (preferredZone !== undefined && selected.zone !== preferredZone) {
      applyTemplateToTile(selected, defaults.templatesByZone[preferredZone]);
      if (preferredZone === "town") {
        addTownLocations(selected, state.radius);
      }
    }

    selected.metadata = {
      ...selected.metadata,
      specialTileId: special.id,
      specialTile: true,
    };
    selected.tags = [...selected.tags, "special-tile"];
    const tileDefinition = getTileById(special.id);
    selected.name = tileDefinition.name;
    if (tileDefinition.enemyIds && tileDefinition.enemyIds.length > 0) {
      selected.enemyPool = tileDefinition.enemyIds.map((enemyId) => ({
        enemyId,
        weight: 1,
      }));
    }
  }
}

type TownPlacementSlot = {
  tile: ExploreTile;
  assignedCount: number;
};

function getTownPlacementSlots(state: ExploreState, spec: WorldSpec): TownPlacementSlot[] {
  const townTiles = state.tiles.filter((tile) => tile.zone === "town");

  for (const tile of townTiles) {
    if (tile.locations.length === 0) {
      addTownLocations(tile, state.radius);
    }
  }

  const slots = townTiles
    .filter((tile) => tile.locations.length > 0)
    .map((tile) => {
      const assignedCount = tile.locations
        .flatMap((location) => location.characters)
        .filter((character) => character.npcId !== undefined).length;

      return {
        tile,
        assignedCount,
      };
    });

  slots.sort((a, b) => hashToUnitInterval(`${spec.seed}:town-slot:${a.tile.key}`) - hashToUnitInterval(`${spec.seed}:town-slot:${b.tile.key}`));
  return slots;
}

function createFallbackTownLocation(tile: ExploreTile, building: ContentNpc["residenceBuilding"]): ExploreTile["locations"][number] {
  const titleByBuilding: Record<ContentNpc["residenceBuilding"], string> = {
    shop: "Provision Shop",
    inn: "Roadside Inn",
    guild: "Guild Hall",
    square: "Town Square",
    shrine: "Wayside Shrine",
  };

  return {
    id: `${tile.key}:generated-${building}`,
    name: titleByBuilding[building],
    description: `A ${building} where local residents gather and exchange news.`,
    locationType: building,
    characters: [],
    tags: ["generated", building],
  };
}

function getOrCreateBuildingLocation(
  tile: ExploreTile,
  npc: ContentNpc,
  spec: WorldSpec,
): ExploreTile["locations"][number] {
  const existing = tile.locations.find((location) => location.locationType === npc.residenceBuilding);
  if (existing !== undefined) {
    return existing;
  }

  const generated = createFallbackTownLocation(tile, npc.residenceBuilding);
  const insertIndex = Math.floor(hashToUnitInterval(`${spec.seed}:insert-location:${tile.key}:${npc.id}`) * (tile.locations.length + 1));
  tile.locations.splice(Math.max(0, Math.min(tile.locations.length, insertIndex)), 0, generated);
  return generated;
}

function assignNpcToTown(slot: TownPlacementSlot, npc: ContentNpc, spec: WorldSpec): boolean {
  const hasNpcAlready = slot.tile.locations.some((location) =>
    location.characters.some((character) => character.npcId === npc.id),
  );

  if (hasNpcAlready || slot.assignedCount >= NPCS_PER_TOWN) {
    return false;
  }

  const targetLocation = getOrCreateBuildingLocation(slot.tile, npc, spec);

  targetLocation.characters.push({
    id: `${slot.tile.key}:npc:${npc.id}`,
    npcId: npc.id,
    name: npc.name,
    role: npc.role,
    disposition: "neutral",
    description: npc.notes,
    questHooks: [],
  });

  slot.assignedCount += 1;
  return true;
}

function placeTownNpcs(state: ExploreState, spec: WorldSpec): void {
  const slots = getTownPlacementSlots(state, spec);
  if (slots.length === 0) {
    return;
  }

  const usedNpcIds: Record<string, true> = {};
  const requiredNpcIds: Record<string, true> = {};

  const requiredNpcs = spec.requiredNpcs.map((requirement) => {
    requiredNpcIds[requirement.npcId] = true;
    return getNpcById(requirement.npcId);
  });

  requiredNpcs.sort((a, b) => hashToUnitInterval(`${spec.seed}:required:${a.id}`) - hashToUnitInterval(`${spec.seed}:required:${b.id}`));

  for (const npc of requiredNpcs) {
    if (usedNpcIds[npc.id] === true) {
      continue;
    }

    const candidates = [...slots].filter((entry) => entry.assignedCount < NPCS_PER_TOWN);
    candidates.sort((a, b) => {
      if (a.assignedCount !== b.assignedCount) {
        return a.assignedCount - b.assignedCount;
      }

      const aBias = hashToUnitInterval(`${spec.seed}:required-town:${npc.id}:${a.tile.key}`);
      const bBias = hashToUnitInterval(`${spec.seed}:required-town:${npc.id}:${b.tile.key}`);
      return aBias - bBias;
    });

    const target = candidates[0];
    if (target === undefined) {
      break;
    }

    if (assignNpcToTown(target, npc, spec)) {
      usedNpcIds[npc.id] = true;
    }
  }

  const ambientPool = listNpcs().filter(
    (npc) => npc.defaultZone === "town" && requiredNpcIds[npc.id] !== true,
  );
  ambientPool.sort((a, b) => hashToUnitInterval(`${spec.seed}:ambient:${a.id}`) - hashToUnitInterval(`${spec.seed}:ambient:${b.id}`));

  for (const npc of ambientPool) {
    if (usedNpcIds[npc.id] === true) {
      continue;
    }

    const candidates = [...slots].filter((entry) => entry.assignedCount < NPCS_PER_TOWN);
    if (candidates.length === 0) {
      break;
    }

    candidates.sort((a, b) => {
      if (a.assignedCount !== b.assignedCount) {
        return a.assignedCount - b.assignedCount;
      }

      const aBias = hashToUnitInterval(`${spec.seed}:ambient-town:${npc.id}:${a.tile.key}`);
      const bBias = hashToUnitInterval(`${spec.seed}:ambient-town:${npc.id}:${b.tile.key}`);
      return aBias - bBias;
    });

    const target = candidates[0];
    if (target === undefined) {
      break;
    }

    if (assignNpcToTown(target, npc, spec)) {
      usedNpcIds[npc.id] = true;
    }
  }

  for (const slot of slots) {
    slot.tile.locations = slot.tile.locations.filter((location) => location.characters.length > 0);
  }
}

function assignExplorationFlows(state: ExploreState, seed: string): void {
  const allFlows = listExplorationFlows();

  const poolByZone: Partial<Record<ZoneType, ContentExplorationFlow[]>> = {};
  for (const flow of allFlows) {
    if (poolByZone[flow.zone] === undefined) {
      poolByZone[flow.zone] = [];
    }
    poolByZone[flow.zone]!.push(flow);
  }

  const candidates = state.tiles
    .filter((tile) => tile.zone !== "town")
    .sort((a, b) => (a.key < b.key ? -1 : a.key > b.key ? 1 : 0));

  for (const tile of candidates) {
    const pool = poolByZone[tile.zone];
    if (pool === undefined || pool.length === 0) {
      continue;
    }

    const score = hashToUnitInterval(`${seed}:flow:${tile.key}`);
    const index = Math.max(0, Math.min(pool.length - 1, Math.floor(score * pool.length)));
    tile.explorationFlowId = pool[index].id;
    pool.splice(index, 1);
  }
}

export function applyWorldSpecToExploreState(state: ExploreState, spec: WorldSpec): void {
  ensureZoneMinimums(state, spec);
  ensureSpecialTiles(state, spec);
  enforceMaxTownTiles(state, spec);
  placeTownNpcs(state, spec);
  assignExplorationFlows(state, spec.seed);
}

File: exploration/world-generator.lua
local ____lualib = require("lualib_bundle")
local __TS__StringCharCodeAt = ____lualib.__TS__StringCharCodeAt
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local __TS__ArraySort = ____lualib.__TS__ArraySort
local __TS__ArrayFind = ____lualib.__TS__ArrayFind
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__SparseArrayNew = ____lualib.__TS__SparseArrayNew
local __TS__SparseArrayPush = ____lualib.__TS__SparseArrayPush
local __TS__SparseArraySpread = ____lualib.__TS__SparseArraySpread
local __TS__ArrayFlatMap = ____lualib.__TS__ArrayFlatMap
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__ArraySome = ____lualib.__TS__ArraySome
local ____exports = {}
local ____tile_2Dfactory = require("exploration.tile-factory")
local createDefaultTileFactoryConfig = ____tile_2Dfactory.createDefaultTileFactoryConfig
local ____explore_2Dstate = require("exploration.explore-state")
local getHexDistance = ____explore_2Dstate.getHexDistance
local ____content_2Dregistry = require("planning.content-registry")
local getNpcById = ____content_2Dregistry.getNpcById
local getTileById = ____content_2Dregistry.getTileById
local listExplorationFlows = ____content_2Dregistry.listExplorationFlows
local listNpcs = ____content_2Dregistry.listNpcs
local START_TILE_KEY = "0,0"
local NPCS_PER_TOWN = 3
local function hashToUnitInterval(self, input)
    local hash = 2166136261
    do
        local i = 0
        while i < #input do
            hash = bit.bxor(
                hash,
                __TS__StringCharCodeAt(input, i)
            )
            hash = hash * 16777619
            i = i + 1
        end
    end
    local normalized = bit.rshift(hash, 0) / 4294967296
    return normalized
end
local function applyTemplateToTile(self, tile, template)
    tile.zone = template.zone
    tile.name = template.name
    tile.description = template.description
    tile.color = {template.color[1], template.color[2], template.color[3], template.color[4]}
    tile.explorationFlowId = nil
    tile.flowLevel = 0
    tile.enemyPool = __TS__ArrayMap(
        template.enemyPool,
        function(____, entry) return {enemyId = entry.enemyId, weight = entry.weight} end
    )
    tile.tags = {unpack(template.tags)}
    tile.locations = {}
end
local function pickCandidateTiles(self, state, targetZone)
    local pool = __TS__ArrayFilter(
        state.tiles,
        function(____, tile) return tile.key ~= START_TILE_KEY and tile.zone ~= targetZone end
    )
    __TS__ArraySort(
        pool,
        function(____, a, b)
            local aScore = hashToUnitInterval(nil, (a.key .. ":") .. targetZone)
            local bScore = hashToUnitInterval(nil, (b.key .. ":") .. targetZone)
            return aScore - bScore
        end
    )
    return pool
end
local function addTownLocations(self, tile, radius)
    local defaults = createDefaultTileFactoryConfig(nil)
    local buildLocations = defaults.buildLocations
    if buildLocations == nil then
        return
    end
    local distanceFromCenter = getHexDistance(nil, {q = 0, r = 0}, tile.coord)
    local locations = buildLocations(
        nil,
        "town",
        {
            key = tile.key,
            coord = {q = tile.coord.q, r = tile.coord.r},
            radius = radius,
            distanceFromCenter = distanceFromCenter,
            roll = hashToUnitInterval(
                nil,
                (("roll:" .. tile.key) .. ":") .. tostring(radius)
            )
        }
    )
    tile.locations = locations
end
local function ensureZoneMinimums(self, state, spec)
    local defaults = createDefaultTileFactoryConfig(nil)
    for ____, requirement in ipairs(spec.requiredZones) do
        do
            local currentCount = #__TS__ArrayFilter(
                state.tiles,
                function(____, tile) return tile.zone == requirement.zone end
            )
            if currentCount >= requirement.minCount then
                goto __continue13
            end
            local needed = requirement.minCount - currentCount
            local candidates = pickCandidateTiles(nil, state, requirement.zone)
            do
                local i = 0
                while i < needed and i < #candidates do
                    local tile = candidates[i + 1]
                    applyTemplateToTile(nil, tile, defaults.templatesByZone[requirement.zone])
                    if requirement.zone == "town" then
                        addTownLocations(nil, tile, state.radius)
                    end
                    i = i + 1
                end
            end
        end
        ::__continue13::
    end
end
local function pickReplacementZone(self, tileKey, seed)
    local pool = {"forest", "farmland", "mountain", "ocean"}
    local index = math.floor(hashToUnitInterval(nil, (seed .. ":replace:") .. tileKey) * #pool)
    return pool[math.max(
        0,
        math.min(#pool - 1, index)
    ) + 1]
end
local function enforceMaxTownTiles(self, state, spec)
    local ____opt_0 = __TS__ArrayFind(
        spec.requiredZones,
        function(____, entry) return entry.zone == "town" end
    )
    local townMinRequirement = ____opt_0 and ____opt_0.minCount or 0
    local maxTownTiles = math.max(spec.maxTownTiles, townMinRequirement)
    local townTiles = __TS__ArrayFilter(
        state.tiles,
        function(____, tile) return tile.zone == "town" end
    )
    if #townTiles <= maxTownTiles then
        return
    end
    local protectedKeys = {[START_TILE_KEY] = true}
    local candidates = __TS__ArrayFilter(
        townTiles,
        function(____, tile) return protectedKeys[tile.key] ~= true end
    )
    __TS__ArraySort(
        candidates,
        function(____, a, b) return hashToUnitInterval(nil, (spec.seed .. ":") .. a.key) - hashToUnitInterval(nil, (spec.seed .. ":") .. b.key) end
    )
    local toConvert = #townTiles - maxTownTiles
    for ____, tile in ipairs(candidates) do
        if toConvert <= 0 then
            break
        end
        local replacementZone = pickReplacementZone(nil, tile.key, spec.seed)
        local defaults = createDefaultTileFactoryConfig(nil)
        applyTemplateToTile(nil, tile, defaults.templatesByZone[replacementZone])
        toConvert = toConvert - 1
    end
end
local function ensureSpecialTiles(self, state, spec)
    local defaults = createDefaultTileFactoryConfig(nil)
    for ____, special in ipairs(spec.requiredSpecialTiles) do
        do
            local existing = __TS__ArrayFilter(
                state.tiles,
                function(____, tile) return tile.metadata.specialTileId == special.id end
            )
            if #existing >= special.minCount then
                goto __continue31
            end
            local preferredZone = special.preferredZone
            local candidatePool = __TS__ArrayFilter(
                state.tiles,
                function(____, tile)
                    if tile.key == START_TILE_KEY then
                        return false
                    end
                    if tile.metadata.specialTileId ~= nil then
                        return false
                    end
                    if preferredZone ~= nil then
                        return tile.zone == preferredZone
                    end
                    return true
                end
            )
            if #candidatePool == 0 then
                goto __continue31
            end
            __TS__ArraySort(
                candidatePool,
                function(____, a, b) return hashToUnitInterval(nil, a.key) - hashToUnitInterval(nil, b.key) end
            )
            local selected = candidatePool[1]
            if preferredZone ~= nil and selected.zone ~= preferredZone then
                applyTemplateToTile(nil, selected, defaults.templatesByZone[preferredZone])
                if preferredZone == "town" then
                    addTownLocations(nil, selected, state.radius)
                end
            end
            selected.metadata = __TS__ObjectAssign({}, selected.metadata, {specialTileId = special.id, specialTile = true})
            local ____array_2 = __TS__SparseArrayNew(unpack(selected.tags))
            __TS__SparseArrayPush(____array_2, "special-tile")
            selected.tags = {__TS__SparseArraySpread(____array_2)}
            local tileDefinition = getTileById(nil, special.id)
            selected.name = tileDefinition.name
            if tileDefinition.enemyIds and #tileDefinition.enemyIds > 0 then
                selected.enemyPool = __TS__ArrayMap(
                    tileDefinition.enemyIds,
                    function(____, enemyId) return {enemyId = enemyId, weight = 1} end
                )
            end
        end
        ::__continue31::
    end
end
local function getTownPlacementSlots(self, state, spec)
    local townTiles = __TS__ArrayFilter(
        state.tiles,
        function(____, tile) return tile.zone == "town" end
    )
    for ____, tile in ipairs(townTiles) do
        if #tile.locations == 0 then
            addTownLocations(nil, tile, state.radius)
        end
    end
    local slots = __TS__ArrayMap(
        __TS__ArrayFilter(
            townTiles,
            function(____, tile) return #tile.locations > 0 end
        ),
        function(____, tile)
            local assignedCount = #__TS__ArrayFilter(
                __TS__ArrayFlatMap(
                    tile.locations,
                    function(____, location) return location.characters end
                ),
                function(____, character) return character.npcId ~= nil end
            )
            return {tile = tile, assignedCount = assignedCount}
        end
    )
    __TS__ArraySort(
        slots,
        function(____, a, b) return hashToUnitInterval(nil, (spec.seed .. ":town-slot:") .. a.tile.key) - hashToUnitInterval(nil, (spec.seed .. ":town-slot:") .. b.tile.key) end
    )
    return slots
end
local function createFallbackTownLocation(self, tile, building)
    local titleByBuilding = {
        shop = "Provision Shop",
        inn = "Roadside Inn",
        guild = "Guild Hall",
        square = "Town Square",
        shrine = "Wayside Shrine"
    }
    return {
        id = (tile.key .. ":generated-") .. building,
        name = titleByBuilding[building],
        description = ("A " .. building) .. " where local residents gather and exchange news.",
        locationType = building,
        characters = {},
        tags = {"generated", building}
    }
end
local function getOrCreateBuildingLocation(self, tile, npc, spec)
    local existing = __TS__ArrayFind(
        tile.locations,
        function(____, location) return location.locationType == npc.residenceBuilding end
    )
    if existing ~= nil then
        return existing
    end
    local generated = createFallbackTownLocation(nil, tile, npc.residenceBuilding)
    local insertIndex = math.floor(hashToUnitInterval(nil, (((spec.seed .. ":insert-location:") .. tile.key) .. ":") .. npc.id) * (#tile.locations + 1))
    __TS__ArraySplice(
        tile.locations,
        math.max(
            0,
            math.min(#tile.locations, insertIndex)
        ),
        0,
        generated
    )
    return generated
end
local function assignNpcToTown(self, slot, npc, spec)
    local hasNpcAlready = __TS__ArraySome(
        slot.tile.locations,
        function(____, location) return __TS__ArraySome(
            location.characters,
            function(____, character) return character.npcId == npc.id end
        ) end
    )
    if hasNpcAlready or slot.assignedCount >= NPCS_PER_TOWN then
        return false
    end
    local targetLocation = getOrCreateBuildingLocation(nil, slot.tile, npc, spec)
    local ____targetLocation_characters_3 = targetLocation.characters
    ____targetLocation_characters_3[#____targetLocation_characters_3 + 1] = {
        id = (slot.tile.key .. ":npc:") .. npc.id,
        npcId = npc.id,
        name = npc.name,
        role = npc.role,
        disposition = "neutral",
        description = npc.notes,
        questHooks = {}
    }
    slot.assignedCount = slot.assignedCount + 1
    return true
end
local function placeTownNpcs(self, state, spec)
    local slots = getTownPlacementSlots(nil, state, spec)
    if #slots == 0 then
        return
    end
    local usedNpcIds = {}
    local requiredNpcIds = {}
    local requiredNpcs = __TS__ArrayMap(
        spec.requiredNpcs,
        function(____, requirement)
            requiredNpcIds[requirement.npcId] = true
            return getNpcById(nil, requirement.npcId)
        end
    )
    __TS__ArraySort(
        requiredNpcs,
        function(____, a, b) return hashToUnitInterval(nil, (spec.seed .. ":required:") .. a.id) - hashToUnitInterval(nil, (spec.seed .. ":required:") .. b.id) end
    )
    for ____, npc in ipairs(requiredNpcs) do
        do
            if usedNpcIds[npc.id] == true then
                goto __continue67
            end
            local candidates = __TS__ArrayFilter(
                {unpack(slots)},
                function(____, entry) return entry.assignedCount < NPCS_PER_TOWN end
            )
            __TS__ArraySort(
                candidates,
                function(____, a, b)
                    if a.assignedCount ~= b.assignedCount then
                        return a.assignedCount - b.assignedCount
                    end
                    local aBias = hashToUnitInterval(nil, (((spec.seed .. ":required-town:") .. npc.id) .. ":") .. a.tile.key)
                    local bBias = hashToUnitInterval(nil, (((spec.seed .. ":required-town:") .. npc.id) .. ":") .. b.tile.key)
                    return aBias - bBias
                end
            )
            local target = candidates[1]
            if target == nil then
                break
            end
            if assignNpcToTown(nil, target, npc, spec) then
                usedNpcIds[npc.id] = true
            end
        end
        ::__continue67::
    end
    local ambientPool = __TS__ArrayFilter(
        listNpcs(nil),
        function(____, npc) return npc.defaultZone == "town" and requiredNpcIds[npc.id] ~= true end
    )
    __TS__ArraySort(
        ambientPool,
        function(____, a, b) return hashToUnitInterval(nil, (spec.seed .. ":ambient:") .. a.id) - hashToUnitInterval(nil, (spec.seed .. ":ambient:") .. b.id) end
    )
    for ____, npc in ipairs(ambientPool) do
        do
            if usedNpcIds[npc.id] == true then
                goto __continue77
            end
            local candidates = __TS__ArrayFilter(
                {unpack(slots)},
                function(____, entry) return entry.assignedCount < NPCS_PER_TOWN end
            )
            if #candidates == 0 then
                break
            end
            __TS__ArraySort(
                candidates,
                function(____, a, b)
                    if a.assignedCount ~= b.assignedCount then
                        return a.assignedCount - b.assignedCount
                    end
                    local aBias = hashToUnitInterval(nil, (((spec.seed .. ":ambient-town:") .. npc.id) .. ":") .. a.tile.key)
                    local bBias = hashToUnitInterval(nil, (((spec.seed .. ":ambient-town:") .. npc.id) .. ":") .. b.tile.key)
                    return aBias - bBias
                end
            )
            local target = candidates[1]
            if target == nil then
                break
            end
            if assignNpcToTown(nil, target, npc, spec) then
                usedNpcIds[npc.id] = true
            end
        end
        ::__continue77::
    end
    for ____, slot in ipairs(slots) do
        slot.tile.locations = __TS__ArrayFilter(
            slot.tile.locations,
            function(____, location) return #location.characters > 0 end
        )
    end
end
local function assignExplorationFlows(self, state, seed)
    local allFlows = listExplorationFlows(nil)
    local poolByZone = {}
    for ____, flow in ipairs(allFlows) do
        if poolByZone[flow.zone] == nil then
            poolByZone[flow.zone] = {}
        end
        local ____poolByZone_flow_zone_4 = poolByZone[flow.zone]
        ____poolByZone_flow_zone_4[#____poolByZone_flow_zone_4 + 1] = flow
    end
    local candidates = __TS__ArraySort(
        __TS__ArrayFilter(
            state.tiles,
            function(____, tile) return tile.zone ~= "town" end
        ),
        function(____, a, b) return a.key < b.key and -1 or (a.key > b.key and 1 or 0) end
    )
    for ____, tile in ipairs(candidates) do
        do
            local pool = poolByZone[tile.zone]
            if pool == nil or #pool == 0 then
                goto __continue95
            end
            local score = hashToUnitInterval(nil, (seed .. ":flow:") .. tile.key)
            local index = math.max(
                0,
                math.min(
                    #pool - 1,
                    math.floor(score * #pool)
                )
            )
            tile.explorationFlowId = pool[index + 1].id
            __TS__ArraySplice(pool, index, 1)
        end
        ::__continue95::
    end
end
function ____exports.applyWorldSpecToExploreState(self, state, spec)
    ensureZoneMinimums(nil, state, spec)
    ensureSpecialTiles(nil, state, spec)
    enforceMaxTownTiles(nil, state, spec)
    placeTownNpcs(nil, state, spec)
    assignExplorationFlows(nil, state, spec.seed)
end
return ____exports
